src/daft-json/src/decoding.rs

```diff
@@ -0,0 +1,534 @@
+        }
+        DataType::UInt8 => Box::new(MutablePrimitiveArray::<u8>::with_capacity(length)),
+        DataType::UInt16 => Box::new(MutablePrimitiveArray::<u16>::with_capacity(length)),
+        DataType::UInt32 => Box::new(MutablePrimitiveArray::<u32>::with_capacity(length)),
+        DataType::UInt64 => Box::new(MutablePrimitiveArray::<u64>::with_capacity(length)),
+        DataType::Float16 => Box::new(MutablePrimitiveArray::<f16>::with_capacity(length)),
+        DataType::Float32 => Box::new(MutablePrimitiveArray::<f32>::with_capacity(length)),
+        DataType::Float64 => Box::new(MutablePrimitiveArray::<f64>::with_capacity(length)),
+        DataType::Boolean => Box::new(MutableBooleanArray::with_capacity(length)),
+        DataType::Utf8 => Box::new(MutableUtf8Array::<i32>::with_capacity(length)),
+        DataType::LargeUtf8 => Box::new(MutableUtf8Array::<i64>::with_capacity(length)),
+        DataType::FixedSizeList(inner, size) => Box::new(MutableFixedSizeListArray::new_from(
+            allocate_array(inner, length),
+            f.data_type().clone(),
+            *size,
+        )),
+        // TODO(Clark): Ensure that these mutable list arrays work correctly and efficiently for arbitrarily nested arrays.
+        // TODO(Clark): We have to manually give a non-None bitmap due to a bug in try_extend_from_lengths for
+        // mutable list arrays, which will unintentionally drop the validity mask if the bitmap isn't already non-None.
+        DataType::List(inner) => Box::new(MutableListArray::new_from_mutable(
+            allocate_array(inner, length),
+            Offsets::<i32>::with_capacity(length),
+            Some(MutableBitmap::with_capacity(length)),
+        )),
+        DataType::LargeList(inner) => Box::new(MutableListArray::new_from_mutable(
+            allocate_array(inner, length),
+            Offsets::<i64>::with_capacity(length),
+            Some(MutableBitmap::with_capacity(length)),
+        )),
+        // TODO(Clark): We have to manually give a non-None bitmap due to a bug in MutableStructArray::push(), which will
+        // unintentionally drop the first null added to the validity mask if a bitmap hasn't been initialized from the start.
+        dt @ DataType::Struct(inner) => Box::new(
+            MutableStructArray::try_new(
+                dt.clone(),
+                inner
+                    .iter()
+                    .map(|field| allocate_array(field, length))
+                    .collect::<Vec<_>>(),
+                Some(MutableBitmap::with_capacity(length)),
+            )
+            .unwrap(),
+        ),
+        dt => todo!("Dtype not supported: {:?}", dt),
+    }
+}
+
+/// Deserialize `rows` by extending them into the given `target`
+fn deserialize_into<'a, A: Borrow<Value<'a>>>(target: &mut Box<dyn MutableArray>, rows: &[A]) {
+    match target.data_type() {
+        DataType::Null => {
+            // TODO(Clark): Return an error if any of rows are not Value::Null.
```

Comment:
is this TODO important?

Reply:
Nah, we still need to add a "strict" parsing mode for both the CSV reader and JSON reader, so this TODO can be addressed when we add that parsing mode; for now, the readers are very forgiving, falling back to UTF8 when parsing fails and trusting the inferred dtype (e.g. here with the `DataType::Null` dtype).